consensus/core/src/commit_finalizer.rs

```diff
@@ -320,87 +323,59 @@ impl CommitFinalizer {
+    // Children of children blocks should not vote on the voted block. And their votes should be ignored.
+    // So after visiting a block when collecting votes, all of the visited block's direct ancestors do not need to be visited.
+    fn append_direct_ancestors_in_last_commit(&mut self) {
         let commit_state = self
             .pending_commits
             .back_mut()
             .unwrap_or_else(|| panic!("No pending commit."));
-
-        // Inherit in ascending order of round, to ensure all lower round reject votes are included.
         let mut blocks = commit_state.commit.blocks.clone();
         blocks.sort_by_key(|b| b.round());
-
+        let blocks_map = self.blocks.read();
         for block in blocks {
-            // Inherit reject votes from the ancestor of block's own authority.
-            // Block verification ensures the 1st ancestor is from the own authority.
-            // If this is not the case, the ancestor block must have been gc'ed.
-            // Also, block verification ensures each authority has at most one ancestor.
-            let Some(own_ancestor) = block.ancestors().first().copied() else {
-                continue;
-            };
-            if own_ancestor.author != block.author() {
-                continue;
-            }
-            let Some(own_ancestor_rejects) = self
-                .blocks
-                .get(&own_ancestor)
-                .map(|b| b.reject_votes.clone())
-            else {
-                // The ancestor block has been finalized or gc'ed.
-                // So its reject votes are no longer needed either.
-                continue;
-            };
-            // No reject votes from the ancestor block to inherit.
-            if own_ancestor_rejects.is_empty() {
-                continue;
-            }
-            // Otherwise, inherit the reject votes.
-            let block_state = self.blocks.get_mut(&block.reference()).unwrap();
-            for (block_ref, reject_votes) in own_ancestor_rejects {
-                block_state
-                    .reject_votes
-                    .entry(block_ref)
-                    .or_default()
-                    .extend(reject_votes.iter());
+            let block_ref = block.reference();
+            // Block verification ensures the first ancestor is from the block's own authority.
+            // Also, block verification ensures each authority appears at most once among ancestors.
+            let mut direct_ancestor = *blocks_map
+                .get(&block_ref)
+                .unwrap()
+                .read()
+                .block
+                .ancestors()
+                .first()
+                .unwrap();
+            while direct_ancestor.author == block_ref.author {
+                let Some(ancestor_block) = blocks_map.get(&direct_ancestor) else {
+                    break;
+                };
+                ancestor_block.write().direct_ancestors.push(block_ref);
```

Comment:
Why `direct_ancestors` ? It looks like we are setting to the `ancestor_block` all the block refs that are actually directly reference the `ancestor_block`. So my understanding is that the `block_ref` has as a direct ancestor the `ancestor_block` . Isn't it the case? Or by `direct_ancestors` we want to signify the blocks that are having the `ancestor_block` as direct ancestor?